source/Test.cpp

```diff
@@ -210,10 +370,36 @@ const string &Test::StatusText() const
 
 
 // Fail the test using the given message as reason.
-void Test::Fail(const string &testFailMessage) const
+void Test::Fail(const Context &context, PlayerInfo &player, const string &testFailReason) const
 {
+	string testFailMessage = "";
+	// Describe the step where the test failed
+	if(context.stepToRun >= 0 && context.stepToRun < steps.size())
+	{
+		const TestStep &stepToRun = steps[context.stepToRun];
+		const string &stepTypeName = STEPTYPE_TO_TEXT.at(stepToRun.stepType);
+		testFailMessage += "Test step " + to_string(context.stepToRun) + " (" + stepTypeName + ") failed";
+	}
+	else
+		testFailMessage += "Test failed";
+	// Add a reason if we got one.
+	if(!testFailReason.empty())
+		testFailMessage += ": " + testFailReason;
+	else
+		testFailMessage += ".";
+	
+	// Add generic logging of the failed test.
+	Files::LogError(testFailMessage);
+	// Log variables at the moment of failure.
+	const map<string, int64_t> &conditions = player.Conditions();
+	if(!conditions.empty())
+		for(const auto &it : conditions)
+			Files::LogError("Condition: \"" + it.first + "\" = " + to_string(it.second));
```

Comment:
I like your idea of printing only the conditions that this test cares about, but that will be quite some work to implement. What I did for now was to only print the conditions that start with `test: `. This will also limit the amount of conditions that will be printed and gives us room to implement your idea in a later PR.

Just printing the first N conditions is very annoying if the condition somebody wants to see is at N+1.

I also changed the code to roll-up to a huge string and then have a single call for the conditions.